backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalHeapDirectoryHolder.java

```diff
@@ -20,17 +20,36 @@ final class LocalHeapDirectoryHolder implements DirectoryHolder {
 
 	private Directory directory;
 
+	private Directory directoryFromPreviousState;
+	private boolean transferred = false;
+
 	LocalHeapDirectoryHolder(LockFactory lockFactory) {
 		this.lockFactory = lockFactory;
 	}
 
+	@Override
+	public void preStart(DirectoryHolder previousDirectoryHolder) {
+		if ( !( previousDirectoryHolder instanceof LocalHeapDirectoryHolder ) ) {
+			return;
+		}
+		directoryFromPreviousState = ( (LocalHeapDirectoryHolder) previousDirectoryHolder ).transfer();
+	}
+
 	@Override
 	public void start() {
-		this.directory = new ByteBuffersDirectory( lockFactory );
+		if ( directoryFromPreviousState != null ) {
+			directory = directoryFromPreviousState;
+			return;
+		}
+
+		directory = new ByteBuffersDirectory( lockFactory );
```

Comment:
I'd just move everything to one method. Call it start or prestart, it shouldn't matter.